From fd9faeef69db9a9025b9ff4772cc78565447ff63 Mon Sep 17 00:00:00 2001
From: Keshav Amburay <the.ridikulus.rat@gmail.com>
Date: Sat, 26 Apr 2014 15:11:18 -0400
Subject: [PATCH] Allow specifying separate default menuentry for each EFI
 architecture in loader.conf

If both are defined in loader.conf, only the value of architecture specific keyword 'default-<arch>' is used
for setting the default menuentry, and the value of the architecture independent keyword 'default' is ignored.
---
 src/efi/gummiboot.c | 12 +++++++++++-
 1 file changed, 11 insertions(+), 1 deletion(-)

diff --git a/src/efi/gummiboot.c b/src/efi/gummiboot.c
index 935a735..88b15bb 100644
--- a/src/efi/gummiboot.c
+++ b/src/efi/gummiboot.c
@@ -1042,6 +1042,8 @@ static VOID config_defaults_load_from_file(Config *config, CHAR8 *content) {
         CHAR8 *line;
         UINTN pos = 0;
         CHAR8 *key, *value;
+        UINTN default_check = 0;
+        char default_string[] = "default-" MACHINE_TYPE_NAME;
 
         line = content;
         while ((line = line_get_key_value(content, &pos, &key, &value))) {
@@ -1055,7 +1057,13 @@ static VOID config_defaults_load_from_file(Config *config, CHAR8 *content) {
                         continue;
                 }
 
-                if (strcmpa((CHAR8 *)"default", key) == 0) {
+                if (strcmpa((CHAR8 *)default_string, key) == 0) {
+                        FreePool(config->entry_default_pattern);
+                        config->entry_default_pattern = stra_to_str(value);
+                        StrLwr(config->entry_default_pattern);
+                        default_check = 1;
+                        continue;
+                } else if (strcmpa((CHAR8 *)"default", key) == 0 && default_check == 0) {
                         FreePool(config->entry_default_pattern);
                         config->entry_default_pattern = stra_to_str(value);
                         StrLwr(config->entry_default_pattern);
@@ -1097,6 +1105,8 @@ static VOID config_defaults_load_from_file(Config *config, CHAR8 *content) {
                         continue;
                 }
         }
+        
+        FreePool(default_string);
 }
 
 static VOID config_entry_add_from_file(Config *config, EFI_HANDLE *device, CHAR16 *file, CHAR8 *content, CHAR16 *loaded_image_path) {
-- 
1.9.2

